arwen/host/execution.go

```diff
@@ -277,90 +270,86 @@ func (host *vmHost) isBuiltinFunctionName(functionName string) bool {
 	return ok
 }
 
-func (host *vmHost) CreateNewContract(input *vmcommon.ContractCreateInput) ([]byte, error) {
-	log.Trace("CreateNewContract", "len(code)", len(input.ContractCode), "metadata", input.ContractCodeMetadata)
-
-	_, blockchain, metering, output, runtime, _ := host.GetContexts()
-
-	// Use all gas initially. In case of successful deployment, the unused gas
-	// will be restored.
-	initialGasProvided := input.GasProvided
-	metering.UseGas(initialGasProvided)
-
-	if runtime.ReadOnly() {
-		return nil, arwen.ErrInvalidCallOnReadOnlyMode
-	}
-
-	runtime.PushState()
-
-	runtime.SetVMInput(&input.VMInput)
-	address, err := blockchain.NewAddress(input.CallerAddr)
-	if err != nil {
-		runtime.PopSetActiveState()
-		return nil, err
-	}
+func (host *vmHost) CreateNewContract(input *vmcommon.ContractCreateInput) (newAddress []byte, err error) {
+	newAddress = nil
+	err = nil
 
-	err = output.Transfer(address, input.CallerAddr, 0, input.CallValue, nil)
-	if err != nil {
-		runtime.PopSetActiveState()
-		return nil, err
-	}
+	defer func() {
+		if err != nil {
+			newAddress = nil
+		}
+	}()
 
-	blockchain.IncreaseNonce(input.CallerAddr)
-	runtime.SetSCAddress(address)
+	_, blockchain, metering, output, runtime, _ := host.GetContexts()
 
 	codeDeployInput := arwen.CodeDeployInput{
 		ContractCode:         input.ContractCode,
 		ContractCodeMetadata: input.ContractCodeMetadata,
-		ContractAddress:      address,
+		ContractAddress:      newAddress,
 	}
-
 	err = metering.DeductInitialGasForIndirectDeployment(codeDeployInput)
 	if err != nil {
-		runtime.PopSetActiveState()
-		return nil, err
+		return
 	}
 
-	idContext := arwen.AddHostContext(host)
-	runtime.PushInstance()
+	// The init() method of the newly deployed contract will be executed using
+	// host.ExecuteOnDestContext(), which must be informed that calling init() is
+	// permitted in this specific case. In all other cases, it is an error to
+	// call init() using host.ExecuteOnDestContext().
+	host.allowCallingInitIndirectly = true
 
-	gasForDeployment := runtime.GetVMInput().GasProvided
-	err = runtime.StartWasmerInstance(input.ContractCode, gasForDeployment)
-	if err != nil {
-		runtime.PopInstance()
-		runtime.PopSetActiveState()
-		arwen.RemoveHostContext(idContext)
-		return nil, err
+	// Async calls are disallowed during indirect contract deployment.
+	// TODO analyze whether this restriction should be lifted, and how to handle
+	// the async calls here (async calls during indirect deployment can be very
+	// useful).
+	host.allowAsyncCalls = false
+
+	defer func() {
+		host.allowCallingInitIndirectly = false
+		host.allowAsyncCalls = true
+	}()
+
+	if runtime.ReadOnly() {
+		err = arwen.ErrInvalidCallOnReadOnlyMode
+		return
 	}
 
-	err = runtime.VerifyContractCode()
+	newAddress, err = blockchain.NewAddress(input.CallerAddr)
```

Comment:
Perhaps rename to `newContractAddress` (to be similar to the name below)?

Reply:
Renamed.